llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -13704,6 +13704,44 @@ static SDValue expandMul(SDNode *N, SelectionDAG &DAG,
   return SDValue();
 }
 
+// Combine vXi32 (mul (and (lshr X, 15), 0x10001), 0xffff) ->
```

Comment:
I'm wondering if we could generalize this slightly.

mul (and X, 0x10001), 0xfffff is simply a sext(trunc (bitcast to i16) to i1).  That operation can be done via a pair of shifts.  At that point, I think the lshr/shl pair fold in this case.

We could also use demanded bits in place of the actual and node.

Reply:
One extra possible follow on - this could probably be a generic combine instead.

Reply:
>  At that point, I think the lshr/shl pair fold in this case.

I don't think they would since there would be a bitcast between them.